velox/exec/tests/CustomJoinTest.cpp

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/JoinBridge.h"
+#include "velox/exec/tests/utils/OperatorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+
+class CustomJoinNode : public core::PlanNode {
+ public:
+  CustomJoinNode(
+      const core::PlanNodeId& id,
+      core::PlanNodePtr left,
+      core::PlanNodePtr right)
+      : PlanNode(id), sources_{std::move(left), std::move(right)} {}
+
+  const RowTypePtr& outputType() const override {
+    return sources_[0]->outputType();
+  }
+
+  const std::vector<core::PlanNodePtr>& sources() const override {
+    return sources_;
+  }
+
+  std::string_view name() const override {
+    return "custom join";
+  }
+
+ private:
+  void addDetails(std::stringstream& /* stream */) const override {}
+
+  std::vector<core::PlanNodePtr> sources_;
+};
+
+class CustomJoinBridge : public JoinBridge {
+ public:
+  void setNumRows(std::optional<int32_t> numRows) {
+    std::vector<ContinuePromise> promises;
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      numRows_ = numRows;
```

Comment:
nit: perhaps, add a check that numRows_ is unset to ensure this method is called only once

```
VELOX_CHECK(!numRows_.hasValue());
```

Reply:
Fixed